cmd/webhook/admission/admission.go

```diff
@@ -0,0 +1,218 @@
+			Result: &metav1.Status{
+				Message: emsg.Error(),
+			},
+		}, nil
+	}
+
+	return &admissionsv1.AdmissionResponse{
+		UID:     req.UID,
+		Allowed: true,
+	}, nil
+}
+
+func Handler(admitter Admitter) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		admitterName := admitter.Name()
+		totalRequests.WithLabelValues(admitterName).Inc()
+
+		if r.Method != "POST" || r.Header.Get("Content-Type") != "application/json" {
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			w.WriteHeader(http.StatusBadRequest)
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Errorf("Unable to read request: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			return
+		}
+
+		review := admissionsv1.AdmissionReview{}
+		err = json.Unmarshal(body, &review)
+		if err != nil {
+			log.Errorf("Unable to parse request: %v", err)
+			w.WriteHeader(http.StatusBadRequest)
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			return
+		}
+
+		operationInfo := fmt.Sprintf(
+			"%s %s %s/%s",
+			review.Request.Operation,
+			review.Request.Kind,
+			review.Request.Namespace,
+			extractName(review.Request),
+		)
+
+		gvr := review.Request.Resource
+		group := gvr.Group
+		if group == "" {
+			group = "zalando.org"
+		}
+
+		labelValues := prometheus.Labels{
+			"admitter":     admitterName,
+			"operation":    string(review.Request.Operation),
+			"group":        group,
+			"version":      gvr.Version,
+			"resource":     gvr.Resource,
+			"sub_resource": review.Request.SubResource,
+		}
+
+		start := time.Now()
+		defer admissionDuration.With(labelValues).
+			Observe(float64(time.Since(start)) / float64(time.Second))
+
+		admResp, err := admitter.Admit(review.Request)
+		if err != nil {
+			log.Errorf("Rejected %s: %v", operationInfo, err)
+			writeResponse(w, errorResponse(review.Request.UID, err))
+			rejectedRequests.With(labelValues).Inc()
+			return
+		}
+
+		log.Infof("Allowed %s", operationInfo)
+		approvedRequests.With(labelValues).Inc()
+		writeResponse(w, admResp)
+	}
+}
+
+func writeResponse(writer http.ResponseWriter, response *admissionsv1.AdmissionResponse) {
+	resp, err := json.Marshal(admissionsv1.AdmissionReview{
+		Response: response,
+	})
+	if err != nil {
+		log.Errorf("unable to serialize response: %v", err)
+		writer.WriteHeader(http.StatusInternalServerError)
+	}
+	if _, err := writer.Write(resp); err != nil {
+		log.Errorf("unable to write response: %v", err)
+	}
+}
+
+func errorResponse(uid types.UID, err error) *admissionsv1.AdmissionResponse {
+	return &admissionsv1.AdmissionResponse{
+		Allowed: false,
+		UID:     uid,
+		Result: &metav1.Status{
+			Message: err.Error(),
+		},
+	}
+}
+
+func extractName(request *admissionsv1.AdmissionRequest) string {
+	if request.Name != "" {
+		return request.Name
+	}
+
+	obj := metav1.PartialObjectMetadata{}
+	if err := json.Unmarshal(request.Object.Raw, &obj); err != nil {
```

Comment:
Question: when parsing the metadata from the raw object, shouldn't we wrap the metadata with the actual object? Do we have a test for the extract name?

Reply:
> when parsing the metadata from the raw object, shouldn't we wrap the metadata with the actual object?
sorry, what do you mean by wrapping the metadata with an actual object?
> Do we have a test for the extract name?
No, but I should add one

Reply:
Added the test

Reply:
> sorry, what do you mean by wrapping the metadata with an actual object?

I was just not aware of the structure of [PartialObjectMetadata](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#PartialObjectMetadata)

:+1: